cantools/database/utils.py

```diff
@@ -86,6 +87,15 @@ def _encode_fields(fields: Sequence[Union["Signal", "Data"]],
 
             unpacked[field.name] = _transform(value)  # type: ignore[operator]
             continue
+        elif type(value) == decimal.Decimal:
+            _transform = decimal.Decimal if field.is_float else round
+            if scaling:
+                if field.offset == 0 and field.scale == 1:
+                    # treat special case to avoid introduction of unnecessary rounding error
+                    unpacked[field.name] = _transform(value)  # type: ignore[operator]
+                    continue
+            unpacked[field.name] = _transform((value - field.decimal.offset) / field.decimal.scale)  # type: ignore[operator]
```

Comment:
If `field` is a `Data` object, this line raises an exception. I think we need an explicit check here

Reply:
Could you please give an example that would raise an exception?

Reply:
```
import decimal
from cantools.database.diagnostics import Did, Data

data = Data(
    name='ExampleData',
    start=0,
    length=8,
)
did = Did(
    identifier=0xFF,
    name='ExampleDid',
    length=1,
    datas=[data]
)

# Will work fine
did.encode(
    {'ExampleData': 123},
    scaling=False
)

# Will raise an exception
did.encode(
    {'ExampleData': decimal.Decimal(123)},
    scaling=False
)
```

Reply:
I get this too, but adding the additional check `isinstance(field, Signal)` gives the following error instead:

```
# Will raise an exception
>>> did.encode(
...     {'ExampleData': decimal.Decimal(123)},
...     scaling=False
... )
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/albin/.python-environments/.venv/lib/python3.11/site-packages/cantools/database/diagnostics/did.py", line 93, in encode
    encoded = encode_data(data,
              ^^^^^^^^^^^^^^^^^
  File "/home/albin/.python-environments/.venv/lib/python3.11/site-packages/cantools/database/utils.py", line 122, in encode_data
    unpacked = _encode_fields(fields, data, scaling)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/albin/.python-environments/.venv/lib/python3.11/site-packages/cantools/database/utils.py", line 109, in _encode_fields
    unpacked[field.name] = field.choice_string_to_number(str(value))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/albin/.python-environments/.venv/lib/python3.11/site-packages/cantools/database/diagnostics/data.py", line 62, in choice_string_to_number
    raise ValueError(f"Data {self.name} has no choices.")
ValueError: Data ExampleData has no choices.

```
By the way, I get the same error with the code in #515.

Reply:
Yes, it highlights that although it's expected that `encode_data` can receive `decimal.Decimal` values to be encoded, `Data` class has no properties that would allow to encode them. I think it would be sensible for `Data` class to have similar implementation of `Decimal` as `Signal`.

Reply:
Thanks for the comment, I added a check for the data type `Signal`. Which leaves the Data class to be handled as before. 
Probably it would be a nice idea to implement `Decimal` in the `Data` class. However I believe that it is out of the scope for this PR.